### PR TITLE
feat(repository): new `repository upgrade` sub-command to validate indices.

### PR DESCRIPTION
**Pull request by aaron-kasten**
This PR adds functionality to validate indices and replaces #2528.

The full feature will add a validation phase to the current V0 to V1 index upgrade command.  This phase will be run before the `commitUpgrade` phase.  If validation fails, the tool will fail with an error indicating the first failed blob(s), leaving the repository locked.  The user may then either rollback and re-run the upgrade, or continue the upgrade (by re-executing `repo upgrade begin` with the `--commit-mode=always` option).  The user may also choose to change the index blob(s) using a Kopia blob-editor tool and then re-validate the index (to check the work) using `repo upgrade validate`.

Re-running `repo upgrade begin` command with the `--commit-mode=always` flag will commit the existing v1 index, delete the v0 index backup, and unlock the repository.

upgrade validate examples:
```
$ KOPIA_UPGRADE_LOCK_ENABLED=1 kopia repo upgrade --upgrade-owner-id jdoe@example.com validate

upgraded index matches old index
```

commit mode "never" example:
```
$ KOPIA_UPGRADE_LOCK_ENABLED=1 kopia repo upgrade --upgrade-owner-id jdoe@example.com begin --commit-mode=never

deleting /Users/aalpar/Library/Caches/kopia/9ba004d5e92006b9/kopia.repository
Repository upgrade lock intent has been placed.
.
.
.
migrating current indices to epoch format
deleting /Users/aalpar/Library/Caches/kopia/9ba004d5e92006b9/kopia.repository
deleting /Users/aalpar/Library/Caches/kopia/9ba004d5e92006b9/kopia.blobcfg
Repository indices have been upgraded.
upgraded index matches old index
WARN Commit mode is set to 'never'.  Skipping commit.

$ KOPIA_UPGRADE_LOCK_ENABLED=1 kopia repo status --upgrade-owner-id jdoe@example.com

Config file:         /Users/jdoe/Library/Application Support/kopia/repository.config
.
.
.
Max pack length:     21 MB
Index Format:        v2

Epoch Manager:       enabled
Current Epoch: 0
.
.
.
Ongoing upgrade:     Upgrading from format version 1 -> 3
Upgrade Time:        2022-11-09 09:56:27.861252 -0800 PST
Upgrade lock:        Locked
Lock status:         Fully Established
```

failed validation and commit mode "always" example:
```
$ KOPIA_UPGRADE_LOCK_ENABLED=1 kopia repo upgrade --upgrade-owner-id jdoe@example.com begin


deleting /Users/aalpar/Library/Caches/kopia/9ba004d5e92006b9/kopia.repository
Repository upgrade lock intent has been placed.
.
.
.
ERROR failed to load index entries for v1 index entry: failed to load index blob with BlobID xn0_081213a775af0bae06d4d81f4f5f0666-sb28e5d344128a4ed-c1: unable to decrypt index blob: error decrypting BLOB xn0_081213a775af0bae06d4d81f4f5f0666-sb28e5d344128a4ed-c1: unable to decrypt content: cipher: message authentication failed

$ KOPIA_UPGRADE_LOCK_ENABLED=1 kopia repo upgrade --upgrade-owner-id jdoe@example.com validate

ERROR failed to load index entries for v1 index entry: failed to load index blob with BlobID xn0_081213a775af0bae06d4d81f4f5f0666-sb28e5d344128a4ed-c1: unable to decrypt index blob: error decrypting BLOB xn0_081213a775af0bae06d4d81f4f5f0666-sb28e5d344128a4ed-c1: unable to decrypt content: cipher: message authentication failed

```

### CODE REVIEW

File: cli/command_repository_upgrade.go [hunk 1]
```diff
@@ -72,9 +86,123 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+
+	err = loadIndexBlobs(ctx, indexEntries, sm, 0, indexBlobInfos)
+	if err != nil {
+		return errors.Wrapf(err, "failed to load index entries for v0 index entry")
+	}
+
+	indexBlobInfos, _, err = sm.IndexReaderV1().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for v1 index")
+	}
```
Comment (redgoat650) [inline cli/command_repository_upgrade.go:141]: Hmm don't the two blob reader versions need to be inputs to this function? Say, assuming some hypothetical future upgrade from version X to version Y, we still want to verify the before-upgrade indices are equivalent to the after-upgrade indices?

**Review by redgoat650 — COMMENTED** [review]
Posed a few questions

File: cli/command_repository_upgrade.go [hunk 2]
```diff
@@ -72,9 +86,123 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+		for _, indexInfo := range indexInfos {
+			assign(indexInfo, index, indexEntries)
+		}
+	}
+	return nil
+}
+
+// validateAction returns an array of strings (report) that describes differences between
+// the V0 index blob and V1 index blob content.  This is used to check that the upgraded index
+// (V1 index) reflects the content of the old V0 index.
```
[the remark this answers is not in the document]
Comment (redgoat650) [reply cli/command_repository_upgrade.go:121]: I see, it was probably copy/paste from #2528

File: cli/command_repository_upgrade.go [hunk 3]
```diff
@@ -72,9 +86,123 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+		log(ctx).Errorf("%v", err)
+		return nil
+	}
+	return err
+}
+
+// checkIndexInfo compare two index infos.  If a mismatch exists, return an error with diagnostic information.
+func checkIndexInfo(i0, i1 index.Info) error {
+	var err error
+	if i0.GetFormatVersion() != i1.GetFormatVersion() {
```
[the remark this answers is not in the document]
Comment (Shrekster) [reply cli/command_repository_upgrade.go:174]: Hmm. @aaron-kasten does `i0.GetFormatVersion()` return us > 1 after upgrade ? Do we need to fix that ?

File: cli/command_repository_upgrade.go [hunk 4]
```diff
@@ -72,9 +86,123 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+	}
+	if i0.GetEncryptionKeyID() != i1.GetEncryptionKeyID() {
+		err = errors.Errorf("mismatched EncryptionKeyIDs: %v %v", i0.GetEncryptionKeyID(), i1.GetEncryptionKeyID())
+	}
+	if i0.GetDeleted() != i1.GetDeleted() {
+		err = errors.Errorf("mismatched Deleted flags: %v %v", i0.GetDeleted(), i1.GetDeleted())
+	}
+	if i0.GetTimestampSeconds() != i1.GetTimestampSeconds() {
+		err = errors.Errorf("mismatched TimestampSeconds: %v %v", i0.GetTimestampSeconds(), i1.GetTimestampSeconds())
+	}
```
[the remark this answers is not in the document]
Comment (aaron-kasten) [reply cli/command_repository_upgrade.go:197]: @redgoat650 moving to just returning one error instead of a collection to simplify implementation.

File: cli/command_repository_upgrade.go (hunk 1 above)
Comment (aaron-kasten) [reply cli/command_repository_upgrade.go:141]: @redgoat650 in theory yes, but that would be designing before we have a use-case.  For now just dealing with v0 -> v1 index upgrade.

File: cli/command_repository_upgrade.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaron-kasten) [reply cli/command_repository_upgrade.go:174]: @redgoat650 @Shrekster This check was inherited from the preexisting index testing.  This should be the blob format version, not the index structure version.

File: cli/command_repository_set_parameters_test.go [hunk 5]
```diff
@@ -182,6 +182,71 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
...
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-cleanup-safety-margin", "10s")
+	env.RunAndExpectFailure(t, "repository", "set-parameters", "--epoch-advance-on-count", "1")
+
+	out = env.RunAndExpectSuccess(t, "repository", "status")
+	require.Contains(t, out, "Epoch Manager:       enabled")
+	require.Contains(t, out, "Index Format:        v2")
+	require.Contains(t, out, "Format version:      3")
+	require.Contains(t, out, "Epoch cleanup margin:    23h0m0s")
+	require.Contains(t, out, "Epoch advance on:        22 blobs or 80.7 MB, minimum 3h0m0s")
+	require.Contains(t, out, "Epoch checkpoint every:  9 epochs")
```
Comment (redgoat650) [inline cli/command_repository_set_parameters_test.go:245]: same suggestion ☝️  for `require.Regexp` would apply here

File: repo/content/committed_read_manager.go [hunk 6]
```diff
@@ -62,6 +62,11 @@ var allIndexBlobPrefixes = []blob.ID{
 	epoch.RangeCheckpointIndexBlobPrefix,
 }
 
+// IndexBlobReader provides and API for reading index blobs.
```
Comment (redgoat650) [inline repo/content/committed_read_manager.go:65]:
```suggestion
// IndexBlobReader provides an API for reading index blobs.
```

File: repo/content/committed_read_manager.go [hunk 7]
```diff
@@ -113,6 +118,26 @@ type SharedManager struct {
...
+
+	return ParseIndexBlob(ibid, d.Bytes(), sm.format)
+}
+
+// IndexReaderV0 return an index reader for reading V0 indexes.
+func (sm *SharedManager) IndexReaderV0() IndexBlobReader {
+	return sm.indexBlobManagerV0
+}
+
+// IndexReaderV1 return an index reader for reading V0 indexes.
```
Comment (redgoat650) [inline repo/content/committed_read_manager.go:136]:
```suggestion
// IndexReaderV1 return an index reader for reading V1 indexes.
```

File: tests/testenv/cli_test_env.go [hunk 8]
```diff
@@ -106,6 +109,39 @@ func (e *CLITest) RunAndExpectSuccess(t *testing.T, args ...string) []string {
...
+	for {
+		// find a location that isn't already
+		// the value we want to write
+		_, err := f.ReadAt(bs[:], i)
+		require.NoError(t, err)
+		if bs[0] != 0x00 {
+			break
+		}
+		i = rand.Int63n(l)
+	}
```
Comment (redgoat650) [inline tests/testenv/cli_test_env.go:139]: Before we enter an infinite loop, maybe we should check the file we're tweaking actually has any non-zero bytes in it?

Alternatively you could change the mechanism of simulating file corruption to XOR a random byte, which works no matter the content of the file (plus you won't need to loop, cause it'll work for any value)

File: cli/command_repository_set_parameters_test.go (hunk 5 above)
Comment (aaron-kasten) [reply cli/command_repository_set_parameters_test.go:245]: @redgoat650 `require.Contains` is part of the testify framework, so changing the implementation of `Contains` will be challenging.  Using, or creating, another assertion that would allow regular expressions is a good approach.  The rest of the code that inspects output from `require.Contains` should be moved over as well (otherwise there will be one bit of code that doesn't conform).

(I added ticket fix this for the kasten branch)

File: repo/format/upgrade_lock.go [hunk 9]
```diff
@@ -86,6 +103,11 @@ func (m *Manager) CommitUpgrade(ctx context.Context) error {
 		return errors.New("no upgrade in progress")
 	}
 
+	// poison V0 index so that old readers won't be able to open it.
+	if err := WriteLegacyIndexPoisonBlob(ctx, m.blobs); err != nil {
+		log(ctx).Errorf("unable to write legacy index poison blob: %v", err)
+	}
+
 	// restore the old format version
```
Comment (Shrekster) [inline repo/format/upgrade_lock.go:111]: Can upgrade succeed without `WriteLegacyIndexPoisonBlob()` succeeding ? What would be the implications for unlocking the repo when that happens ?

Comment (Shrekster) [reply repo/format/upgrade_lock.go:111]: Never mind, this was also the existing behavior. So, you are not changing it.

File: repo/content/index/info.go [hunk 10]
```diff
@@ -92,3 +93,32 @@ func ToInfoStruct(i Info) *InfoStruct {
...
+	if i0.GetEncryptionKeyID() != i1.GetEncryptionKeyID() {
+		qs = append(qs, fmt.Sprintf("mixed EncryptionKeyIDs: %v %v", i0.GetEncryptionKeyID(), i1.GetEncryptionKeyID()))
+	}
+	if i0.GetDeleted() != i1.GetDeleted() {
+		qs = append(qs, fmt.Sprintf("mixed Deleted flags: %v %v", i0.GetDeleted(), i1.GetDeleted()))
+	}
+	if i0.GetTimestampSeconds() != i1.GetTimestampSeconds() {
+		qs = append(qs, fmt.Sprintf("mixed TimestampSeconds: %v %v", i0.GetTimestampSeconds(), i1.GetTimestampSeconds()))
+	}
+	return qs
```
[the remark this answers is not in the document]
Comment (aaron-kasten) [reply repo/content/index/info.go]: @adowair the above was taken out in an earlier version.  There was a bad merge that re-introduced it.

The code has been removed because its more sophisticated that what was required for the use-case.

File: cli/command_repository_set_parameters.go [hunk 11]
```diff
@@ -140,32 +210,28 @@ func (c *commandRepositorySetParameters) run(ctx context.Context, rep repo.Direc
...
-			mp.EpochParameters = epoch.DefaultParameters()
-			upgradeToEpochManager = true
-			mp.IndexVersion = 2
-		}
+	c.setSizeMBParameter(ctx, c.maxPackSizeMB, "maximum pack size", &mp.MaxPackSize, &anyChange)
 
-		if mp.Version < format.FormatVersion2 {
-			mp.Version = format.FormatVersion2
+	// prevent downgrade of index format
+	if c.indexFormatVersion != 0 && c.indexFormatVersion != mp.IndexVersion {
```
Comment (Shrekster) [inline cli/command_repository_set_parameters.go:223]: Might be a good idea to separate this change out in its own PR for better visibility.

cc @jkowalski , I believe the only scenario anyone would ever consider downgrading index version (outside of tests) is during a corrupted repository recovery attempts. Any concerns over disallowing this anymore ?

File: cli/command_repository_set_parameters_test.go [hunk 12]
```diff
@@ -182,6 +182,71 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
 	env.RunAndExpectSuccess(t, "index", "epoch", "list")
 }
 
+func (s *formatSpecificTestSuite) TestRepositorySetParametersDowngrade(t *testing.T) {
```
Comment (Shrekster) [inline cli/command_repository_set_parameters_test.go:185]: This sounds like a negative test which would attempt to perform an invalid index downgrade and successfully hit a failure. Is that correct ? If so then shouldn't we be using `--index-version` operation here somewhere ?

File: cli/command_repository_set_parameters_test.go [hunk 13]
```diff
@@ -182,6 +182,71 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
...
+
+		// You can only upgrade when you are not already upgraded
+		if s.formatVersion < format.MaxFormatVersion {
+			env.RunAndExpectSuccess(t, cmd...)
+		} else {
+			env.RunAndExpectFailure(t, cmd...)
+		}
+	}
+
+	env.RunAndExpectSuccess(t, "repository", "set-parameters", "--upgrade")
```
Comment (Shrekster) [inline cli/command_repository_set_parameters_test.go:226]: nit; Might be a good idea to write a comment here explaining what is the intention of running the `set-parameters --upgrade` command after running the `upgrade` command above ? I think I know what you might be doing here, but it may not be apparent for the readers.

File: cli/command_repository_upgrade.go [hunk 14]
```diff
@@ -72,9 +83,130 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+
+		for _, indexInfo := range indexInfos {
+			assign(indexInfo, which, indexEntries)
+		}
+	}
+
+	return nil
+}
+
+// validateAction returns an error of the new V1 index blob content does not match the source V0 index blob content.
```
Comment (Shrekster) [inline cli/command_repository_upgrade.go:120]:
```suggestion
// validateAction returns an error if the new V1 index blob content does not match the source V0 index blob content.
```

File: cli/command_repository_upgrade.go [hunk 15]
```diff
@@ -72,9 +83,130 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+		return errors.Wrapf(err, "failed to list index blobs for old index")
+	}
+
+	indexBlobInfos1, _, err := sm.IndexReaderV1().ListIndexBlobInfos(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list index blobs for new index")
+	}
+
+	if len(indexBlobInfos0) == 0 && len(indexBlobInfos1) > 0 {
+		log(ctx).Infof("index has already been upgraded.  Nothing to do")
```
Comment (Shrekster) [inline cli/command_repository_upgrade.go:138]:
```suggestion
		log(ctx).Infof("index has already been upgraded, nothing to do")
```

File: cli/command_repository_upgrade.go [hunk 16]
```diff
@@ -72,9 +83,130 @@ func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent command
...
+	for contentID, indexEntryPairs := range indexEntries {
+		if indexEntryPairs[0] == nil || indexEntryPairs[1] == nil {
+			err = errors.Errorf("lop-sided index entries for contentID %q", contentID)
+		} else {
+			err = checkIndexInfo(indexEntryPairs[0], indexEntryPairs[1])
+		}
+
+		if err != nil {
+			break
+		}
```
Comment (Shrekster) [inline cli/command_repository_upgrade.go:161]: It might be a good idea to dump a minimum number of failures before breaking out. This will help answering questions like "Did we somehow spoil just a single index or more of them ?".

You can add a CLI flag to let the user specify a non-default minimum too. Not required though.

File: cli/command_repository_upgrade.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shrekster) [reply cli/command_repository_upgrade.go]: Ah! Thanks for the response.

**Comment by Shrekster** [comment]
Thanks for taking in the feedbacks. I'll take a final pass on this today as soon as I get out of meetings.

File: cli/command_repository_set_parameters_test.go [hunk 17]
```diff
@@ -182,6 +182,49 @@ func (s *formatSpecificTestSuite) TestRepositorySetParametersUpgrade(t *testing.
...
+			require.Contains(t, out, "Format version:      1")
+			require.Contains(t, out, "Epoch Manager:       disabled")
+			env.RunAndExpectFailure(t, "index", "epoch", "list")
+		case format.FormatVersion2:
+			require.Contains(t, out, "Format version:      2")
+			require.Contains(t, out, "Epoch Manager:       enabled")
+			env.RunAndExpectSuccess(t, "index", "epoch", "list")
+		default:
+			require.Contains(t, out, "Format version:      3")
+			require.Contains(t, out, "Epoch Manager:       enabled")
```
Comment (redgoat650) [inline cli/command_repository_set_parameters_test.go:209]: @aaron-kasten - I wasn't suggesting changing the behavior of `require.Contains`. I was suggesting you instead use  the existing `require.Regexp` instead. Both are part of the testify framework.

https://github.com/stretchr/testify/blob/1333b5d3bda8cf5aedcf3e1aaa95cac28aaab892/require/require.go#L1731

File: cli/command_repository_set_parameters_test.go (hunk 5 above)
Comment (redgoat650) [reply cli/command_repository_set_parameters_test.go]: @aaron-kasten - I wasn't suggesting changing the behavior of `require.Contains`. I was suggesting you instead use  the existing `require.Regexp` instead. Both are part of the testify framework.

https://github.com/stretchr/testify/blob/1333b5d3bda8cf5aedcf3e1aaa95cac28aaab892/require/require.go#L173

**Review by Shrekster — APPROVED** [review]
Nice work! The negative test enhancements and JSON syntax support for failure dump is future work.

File: Makefile [hunk 18]
```diff
@@ -486,3 +486,4 @@ perf-benchmark-results:
 	gcloud compute scp $(PERF_BENCHMARK_INSTANCE):psrecord-* tests/perf_benchmark --zone=$(PERF_BENCHMARK_INSTANCE_ZONE) 
 	gcloud compute scp $(PERF_BENCHMARK_INSTANCE):repo-size-* tests/perf_benchmark --zone=$(PERF_BENCHMARK_INSTANCE_ZONE)
 	(cd tests/perf_benchmark && go run process_results.go)
+
```
Comment (Shrekster) [inline Makefile:489]: needless blank line

Comment (Shrekster) [inline Makefile:489]:
```suggestion
```

**Review by Shrekster — APPROVED** [review]
just one minor comment

good attempt at untangling the codecov issues
